Added TwoPhaseNCGPartialPressureFluidProperties

joshuahansel opened the pull request:
This is a step towards #28335.

- Improved doc for `IdealRealGasMixtureFluidProperties` (our basically only mixture model)
- Added generic methods to `VaporMixtureFluidProperties` for converting from molar fractions to mass fractions and getting the mixture molar mass.
- Added a new `TwoPhaseNCGFluidProperties`: `TwoPhaseNCGPartialPressureFluidProperties`, which takes the name of the `TwoPhaseFluidProperties` and then the name of a single `SinglePhaseFluidProperties` mixed into it.
- Added a new option to `FluidPropertiesInterrogator` specifically for `TwoPhaseNCGPartialPressureFluidProperties` that takes `p` (taken to be mixture pressure) and `T` and gives the saturation conditions at that state, including the NCG mass fraction.

joshuahansel committed:
Improved doc for IdealRealGasMixtureFluidProperties

Refs #28335

joshuahansel committed:
Added generic methods to VaporMixtureFluidProperties

- mixtureMolarMass(molar_fractions, molar_masses)
- massFractionsFromMolarFractions(molar_fractions, molar_masses)

Refs #28335

joshuahansel committed:
Added TwoPhaseNCGPartialPressureFluidProperties

Refs #28335

joshuahansel committed:
Added saturation mass fraction option to FluidPropertiesInterrogator

Refs #28335

joshuahansel commented:
@GiudGiud Regarding that fluid properties function I discussed with you, that's in `TwoPhaseNCGPartialPressureFunction`. Based on potential complications with functors, I decided to just make it function-based, enabling it for `FunctionIC` only, not aux kernels (for output purposes). It might be possible to put the bulk of that function into some kind of interface that gets shared with a new aux kernel.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2389317/) on 9aaadd8 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28394/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2389363/) on 9aaadd8 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>90745b</th>
      <th colspan=4>#28394 9aaadd</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.96%</td>
      <td>85.09%</td>
      <td>+0.14%</td>
      <td>94.51%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7641</td>
      <td>7729</td>
      <td>+88</td>
      <td>86</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1353</td>
      <td>1354</td>
      <td>+1</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28394/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28394/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28394/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on modules/fluid_properties/src/fluidproperties/TwoPhaseNCGPartialPressureFluidProperties.C:
```diff
@@ -0,0 +1,59 @@
+
+    _fp_ncg(getUserObject<SinglePhaseFluidProperties>("fp_ncg"))
+{
+  _fp_2phase = &_fe_problem.getUserObject<TwoPhaseFluidProperties>(_2phase_name);
+  _fp_vapor_primary = &_fe_problem.getUserObject<SinglePhaseFluidProperties>(getVaporName());
```
why not move these up in the initializer's list?

GiudGiud replied:
then you could const them

joshuahansel replied:
There's some dumb object creation ordering issue I ran into. I'll investigate.

joshuahansel replied:
Ok here's what it is. The base class decided that these are `const T *` pointers because there is at least one child that creates these things in the constructor, e.g., `IAPWS95TwoPhaseNCGIdealRealFluidProperties`. Note that I want to deprecate that child, so later your suggestion should be able to be done.

GiudGiud commented on modules/fluid_properties/src/fluidproperties/VaporMixtureFluidProperties.C:
```diff
@@ -33,4 +33,30 @@ VaporMixtureFluidProperties::primaryMassFraction(const std::vector<Real> & x) co
+  mooseAssert(molar_fractions.size() == molar_masses.size(), "Must have same size");
+
+  Real molar_mass_mix = 0;
+  for (unsigned int i = 0; i < molar_fractions.size(); i++)
```
```suggestion
  for (const auto i : index_range(molar_fractions))
```

GiudGiud commented on modules/fluid_properties/src/fluidproperties/VaporMixtureFluidProperties.C:
```diff
@@ -33,4 +33,30 @@ VaporMixtureFluidProperties::primaryMassFraction(const std::vector<Real> & x) co
+  const auto molar_mass_mix = mixtureMolarMass(molar_fractions, molar_masses);
+
+  std::vector<Real> mass_fractions(molar_fractions.size(), 0.0);
+  for (unsigned int i = 0; i < molar_fractions.size(); i++)
```
```suggestion
  for (const auto i : index_range(molar_fractions))
```

joshuahansel commented on modules/fluid_properties/src/functions/TwoPhaseNCGPartialPressureFunction.C:
```diff
@@ -0,0 +1,95 @@
+unsigned int
+TwoPhaseNCGPartialPressureFunction::getNumberOfProvidedArguments() const
+{
+  if (isParamSetByUser("arg1"))
+  {
+    if (isParamSetByUser("arg2"))
+      return 2;
+    else
+      return 1;
+  }
+  else
+    return 0;
```
Yeah, the intention is that for example it's an error for a one-arg function for the user specify arg2 but not arg1. I'll reword your suggestion to the error message.

joshuahansel commented on modules/fluid_properties/doc/content/source/fluidproperties/IdealRealGasMixtureFluidProperties.md:
```diff
@@ -1,26 +1,131 @@
+\xi_i = \frac{\psi_i M_i}{M} \,.
+
+Note the following relation for all mass-specific quantities, such as
+specific volume and specific internal energy:
```
I wasn't sure if there was an official name for this approximation in literature. I think Matthias described it as an ideal gas approximation for real gases, so I'll put something to this effect.

GiudGiud replied:
sounds good.
make sure to use a meaningful commit message pls

joshuahansel replied:
> make sure to use a meaningful commit message pls

Are you just saying you don't like "Addressed review" (plus ticket ref.)  or something like that? I need to be more specific?

GiudGiud replied:
if using "Apply suggestion from code review" just add ": typos" or "constness" or ...

GiudGiud replied:
this is not really in the official workflow yet. I think it makes sense though

joshuahansel committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

joshuahansel committed:
Addressed review

- Mentioned ideal gas mixture in doc
- Clarified arg error message in function
- Fixed hidden Function::value

Refs #28335
